### PR TITLE
Add support for reading ROOT histograms

### PR DESCRIPTION
**Pull request by oschulz**

### CODE REVIEW

**Comment by oschulz** [comment]
CC @gipert

**Comment by gipert** [comment]
Sorry for the delay Oli, haven't got time until now. I was thinking about this: it's a pity that `StatsBase.Histogram` does not support under/over-flow bins – it's something we are losing when converting from uproot hists. Do you think it would make sense to extend `Histogram` here to include this feature? Do you have brighter ideas?

File: src/pyjlconv.jl [hunk 1]
```diff
@@ -39,6 +39,25 @@ function awkwardobjectarray2jl(x::PyObject)
...
+    end
+end
+
+_hist_edges(edges::AbstractVector{<:Real}) = _conv_hist_edge(edges)
+_hist_edges(edges::NTuple{N,AbstractVector{<:Real}}) where N = map(_conv_hist_edge, edges)
+
+function roothist2jl(x::PyObject)
+    edges = _hist_edges(x.edges)
+    weights = x.values
+    Histogram(edges, weights)
```
Comment (gipert) [inline src/pyjlconv.jl]: I know it's the default, but we could be over-precautious here and add `closed=:left` here, since ROOT defaults to this?

Comment (oschulz) [reply src/pyjlconv.jl]: Good point - I don't think the StatsBase default will change, but it certainly couldn't hurt.

**Comment by oschulz** [comment]
> it's a pity that StatsBase.Histogram does not support under/over-flow bins – it's something we are losing when converting from uproot hists. Do you think it would make sense to extend Histogram here to include this feature? Do you have brighter ideas?

I've never used ROOTs over- und underflow bins much, personally, but it is a loss of information. We'd have to create a custom histogram class though - certainly possible, but would be more than a few lines of code. Maybe defer until a later time?

**Comment by gipert** [comment]
> Maybe defer until a later time?

Yes, it's not urgent, had to make use of under/over-flow bins only couple of times.

Another thing: do you know if there is any way to know whether a ROOT histogram represents a density? Would be cool to be able to set `Histogram.isdensity`. I will investigate

**Comment by gipert** [comment]
So there is an interesting bit that could be used to set `Histogram.isdensity`: [`kIsAverage`](https://root.cern.ch/doc/v622/classTH1.html#a4c624d3fc881c5290c39814d7f6161d7a60467ac65df036faf357108ebc90af48). I'm not sure I can get that with uproot tho...

**Comment by gipert** [comment]
Indeed: https://github.com/scikit-hep/uproot3-methods/issues/99

**Comment by gipert** [comment]
From what concerns myself you can merge this and tag a new version, once #18 is in!

**Comment by oschulz** [comment]
> So there is an interesting bit that could be used to set Histogram.isdensity: kIsAverage. I'm not sure I can get that with uproot tho...

Do you know the semantics of that? `StatsBase.Histogram` has several normalization modes (mainly density-normalization and full normalization), both take non-homogenous binning into account. I don't know that ROOT's `kIsAverage` does, exactly.

**Comment by oschulz** [comment]
> you can merge this and tag a new version, once #18 is in!

Want to try your hand at that `kIsAverage` status bit first (from what I understand, we can get it?) or should we do that later?

**Comment by gipert** [comment]
It's impossible to get it with uproot3 :(

**Comment by oschulz** [comment]
> It's impossible to get it with uproot3 :(

**Comment by oschulz** [comment]
> It's impossible to get it with uproot3 :(

Ah, sorry, I misunderstood! Ok, I'll merge this, then. :-)

**Comment by gipert** [comment]
> Do you know the semantics of that? StatsBase.Histogram has several normalization modes (mainly density-normalization and full normalization), both take non-homogenous binning into account. I don't know that ROOT's kIsAverage does, exactly.

In my understanding a ROOT histogram with `kIsAverage` is interpreted as a density. When you `TH1:Add` two of them the bin contents are interpreted as counts normalized by the bin width (I'm not 100% percent tho, I should dig into the code).

Looks like uproot3 is not so clever and does not save this info in the `TH1` python object... uproot4 does! I suggest that we leave these two things in the long-term todo list:

- [ ] Implement custom `Histogram` object to store under/over-flow bins
- [ ] Use `kIsAverage` to set `Histogram.isdensity`, once we switch to uproot4

**Comment by oschulz** [comment]
https://github.com/JuliaRegistries/General/pull/38141

**Comment by oschulz** [comment]
> I suggest that we leave these two things in the long-term todo list

Yep - can you open an issue for that, so we don't forget?

**Pull request merged**